### DISCUSSION TITLE
Automatically configured prefix length?

### DISCUSSION THREAD

**Discussion by oskarwaglund** [rsmarples/dhcpcd#126] [Q&A] [answered]
Hi,

I have set up an Ubuntu DHCPv6 server manually with the ISC DHCPv6 server app along with radvd which sends out router advertisements with prefix length 64 for the same IPv6 subnet. I was hoping that dhcpcd on another Ubuntu server would catch on to the router advertisements and configure the received IP address from the DHCP server as a /64 address, but I still get 128.

Do I manually have to add this behavior in some sort of hook, or can dhcpcd be configured to set the correct prefix length on the IP address? I may be missing some fundamental aspect of how DHCPv6 is supposed to work. In the log below you can see that dhcpcd gets the router advertisement with the prefix /64, but that is not used it seems. 

**dhcpd.conf (DHCPv6 server)**
default-lease-time 600;
max-lease-time 7200;

subnet6 fd00::/64 {
  range6 fd00::/124;
}

 **radvd.conf**
 interface enp0s8
{
AdvSendAdvert on;
AdvManagedFlag on;
MinRtrAdvInterval 30;
MaxRtrAdvInterval 100;
prefix fd00::/64
{
AdvOnLink on;
AdvAutonomous off;
AdvRouterAddr off;
};
};

**dhcpcd output**
dhcpcd-7.1.0 starting
enp0s8: executing `/lib/dhcpcd/dhcpcd-run-hooks' PREINIT
enp0s8: executing `/lib/dhcpcd/dhcpcd-run-hooks' CARRIER
DUID 00:04:6c:95:f7:a7:0d:a3:37:4a:8c:75:f2:60:6a:0f:47:3b
enp0s8: IAID 27:19:ad:24
enp0s8: delaying IPv6 router solicitation for 0.7 seconds
enp0s8: soliciting an IPv6 router
enp0s8: sending Router Solicitation
enp0s8: Router Advertisement from fe80::a00:27ff:feaf:3b5a
enp0s8: adding route to fd00::/64
enp0s8: adding default route via fe80::a00:27ff:feaf:3b5a
enp0s8: executing `/lib/dhcpcd/dhcpcd-run-hooks' ROUTERADVERT
enp0s8: reading lease `/var/lib/dhcpcd/enp0s8.lease6'
enp0s8: soliciting a DHCPv6 lease
enp0s8: delaying SOLICIT6 (xid 0x45ecb3), next in 0.7 seconds
enp0s8: broadcasting SOLICIT6 (xid 0x45ecb3), next in 1.1 seconds
enp0s8: ADV fd00::e/128 from fe80::a00:27ff:feaf:3b5a
enp0s8: broadcasting REQUEST6 (xid 0x4f50bd), next in 1.0 seconds
enp0s8: REPLY6 received from fe80::a00:27ff:feaf:3b5a
enp0s8: adding address fd00::e/128
enp0s8: pltime 375 seconds, vltime 600 seconds
enp0s8: renew in 187, rebind in 300, expire in 600 seconds
enp0s8: writing lease `/var/lib/dhcpcd/enp0s8.lease6'
enp0s8: waiting for DHCPv6 DAD to complete
enp0s8: DHCPv6 DAD completed
enp0s8: executing `/lib/dhcpcd/dhcpcd-run-hooks' BOUND6
forking to background
forked to background, child pid 1032
enp0s8: sending NA for fd00::e/128

**ifconfig output (fd00::e is the received IP address from the DHCPv6 server)**
enp0s8: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet6 fd00::e  prefixlen 128  scopeid 0x0<global>
        inet6 fe80::a00:27ff:fe19:ad24  prefixlen 64  scopeid 0x20<link>
        ether 08:00:27:19:ad:24  txqueuelen 1000  (Ethernet)
        RX packets 21  bytes 2206 (2.2 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 53  bytes 5698 (5.6 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

**Comment by rsmarples** [comment, marked as answer]
RA gives you routes based on a prefix. You could optionally derive an automated address from the prefix.
DHCPv6 gives you an address - there is no prefix.

When we add an address, the kernel will automatically add a route if the prefix length is not at the maximal value.
This may conflict with routes the RA wants to add.
Likewise if a route associated with the address expires, the address would be removed when a different route could be used.
As such all DHCPv6 addresses will always have a /128 prefix length.

You've only said what your problem is, not your use case. If you really want a /64 prefix from DHCPv6 then you want to configure for DHCPv6 Prefix Delegation which requires some setup at both server and client. It doesn't work out of the box.
